components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v2/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v2/impl/core/utils/RecoveryUtil.java

```diff
@@ -0,0 +1,342 @@
+
+        // Build next API calls.
+        ArrayList<APICall> apiCallsArrayList = new ArrayList<>();
+        apiCallsArrayList.add(RecoveryUtil
+                .buildApiCall(APICalls.RESET_PASSWORD_API.getType(), Constants.RelationStates.NEXT_REL,
+                        buildURIForBody(tenantDomain, APICalls.RESET_PASSWORD_API.getApiUrl(),
+                                Constants.ACCOUNT_RECOVERY_ENDPOINT_BASEPATH), null));
+        RetryErrorResponse retryErrorResponse = buildRetryErrorResponse(
+                Constants.STATUS_PRECONDITION_FAILED_MESSAGE_DEFAULT, code, description, resetCode, correlationId,
+                apiCallsArrayList);
+        if (log.isDebugEnabled()) {
+            log.debug(description);
+        }
+        return new PreconditionFailedException(retryErrorResponse);
+    }
+
+    /**
+     * Prepend the operation scenario to the existing exception error code.
+     * (Eg: USR-20045)
+     *
+     * @param exceptionErrorCode Existing error code.
+     * @param scenario           Operation scenario.
+     * @return New error code with the scenario prepended.
+     */
+    public static String prependOperationScenarioToErrorCode(String exceptionErrorCode, String scenario) {
+
+        if (StringUtils.isNotEmpty(exceptionErrorCode)) {
+            if (exceptionErrorCode.contains(IdentityRecoveryConstants.EXCEPTION_SCENARIO_SEPARATOR)) {
+                return exceptionErrorCode;
+            }
+            if (StringUtils.isNotEmpty(scenario)) {
+                exceptionErrorCode =
+                        scenario + IdentityRecoveryConstants.EXCEPTION_SCENARIO_SEPARATOR + exceptionErrorCode;
+            }
+        }
+        return exceptionErrorCode;
+    }
+
+    /**
+     * Builds the API context on whether the tenant qualified url is enabled or not. In tenant qualified mode the
+     * ServiceURLBuilder appends the tenant domain to the URI as a path param automatically. But
+     * in non tenant qualified mode we need to append the tenant domain to the path manually.
+     *
+     * @param endpoint Relative endpoint path.
+     * @return Context of the API.
+     */
+    private static String getContext(String tenantDomain, String endpoint, String baseUrl) {
+
+        String context;
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+            context = baseUrl + endpoint;
+        } else {
+            context = String.format(TENANT_CONTEXT_PATH_COMPONENT, tenantDomain) + baseUrl + endpoint;
+        }
+        return context;
+    }
+
+    /**
+     * Builds error response.
+     *
+     * @param errorCode Error code.
+     * @param errorMessage Error message.
+     * @param errorDescription Error description.
+     * @return ErrorResponse.
+     */
+    private static ErrorResponse buildErrorResponse(String errorCode, String errorMessage, String errorDescription) {
+
+        ErrorResponse errorResponse = getErrorBuilder(errorCode, errorMessage, errorDescription).build(log,
```

Comment:
do the return here itself

Reply:
Fixed with 8a5d4f0274d4bc026253b69400b0714ad6acf7a7